File: implementation/TrainNet.py
import pandas as pd
import projNet
import numpy as np
from NetDefinition import *


def oneHotEncode(vector,numClasses):
    #print("One Hot Encode\n")
    encoded_array = np.zeros((vector.size, numClasses), dtype=int)
    encoded_array[np.arange(vector.size),vector] = 1 
    return encoded_array

def testAccuracy(model, test_df):
    # print("In Test Accuracy\n")
    
    
    # print("convert to numpy")
    truth_subset = test_df.iloc[:,0].to_numpy()
    # print(truth_subset)
    test_subset = test_df.iloc[:,1:].to_numpy()/255
    print(test_subset.shape)
    
    # print("Get shape")
    max_examples = test_subset.shape[0]
    # print(max_examples)
    
    index = 0
    totalCorrect = 0
    # print("ready for loop")
    
    while index < max_examples:
        predictions = model.forward(test_subset[index])
        # print("Predicted Value: " +  str(np.argmax(predictions)))
        # print("Actual: " + str(truth_subset[index]))
        if(np.argmax(predictions) == truth_subset[index]):
            totalCorrect+=1
        index+=1
        
    print("Test accuracy: " + str(totalCorrect/max_examples))
        


def main(learningRate, numThreads, epoch, batch_size, trainDataPath,testDataPath, model):
    
    train_df = pd.read_csv(trainDataPath,header=None,skiprows=1)
    t_df = pd.read_csv(testDataPath,header=None,skiprows=1)
    # print(df)
   
    # print(test_df.shape)
    #print("opened data")
    
    model.neuralNet.setLearningRate(learningRate)
    model.neuralNet.setThreadNum(numThreads)

    #print("Set initial values")
    
    for e in range(0,epoch):
        print("epoch: " + str(e))
        #train_df = df.sample(frac=1)
        #test_df = t_df.sample(frac=.2)
        
        # print("\nDimensions\n")
        
        # model.neuralNet.printDim(4)
        
        # print("\weight\n")
        # model.neuralNet.printWeights(4)
        
        # print("\bias\n")
        # model.neuralNet.printBias(4)
        
        
        truth_subset = train_df.iloc[:,0].to_numpy()
        train_subset = train_df.iloc[:,1:].to_numpy()/255
        encoded_truth = oneHotEncode(truth_subset,10)
        
        
        max_examples = encoded_truth.shape[0]
        #print("Max Sample size: " + str(max_examples))
        #print("Max training: " + str(train_subset.shape[0]))
        #print(train_subset)
        #print(encoded_truth)
        index = 0
        avgLoss = 0
        #print("ready for loop")
        while index < max_examples:
            #print(index)

            model.neuralNet.zeroGrad()
            
            # print("Weight Gradients")
            # model.neuralNet.printGrad()
            
            #print("After grad")
            counter = 0
            while counter < batch_size:
                if index >= max_examples:
                    break
                #print(counter)
                model.neuralNet.setInput(train_subset[index])
                predictions = model.forward(train_subset[index])
                #print(predictions)
                #print("Encoded Truth: " + str(encoded_truth[index]))
                loss = model.neuralNet.crossEntropy(predictions, encoded_truth[index])
                avgLoss += loss
                model.neuralNet.backwardStep(predictions,encoded_truth[index])
                index+=1
                counter+=1
                
                # print("Weight Gradients")
                # model.neuralNet.printGrad()
                
                # print("Bias Gradients")
                # model.neuralNet.printBiasGrad()
            #print("finished batch")
            model.neuralNet.updateWeights()
        print("Avg Loss: " + str(avgLoss/max_examples) + " for epoch: " + str(e))
        #testAccuracy(model,test_df)
    print("finished")
    exit(1)
            
        
        
        
    
    

if __name__ == "__main__":
    
    learningRate = 0.002
    numThreads = 8
    epoch = 50
    batch = 200
    
    trainingDataPath = "../data/mnist_train.csv"
    testingDataPath = "../data/mnist_test.csv"
    
    model = modelMLP()
    
    main(learningRate, numThreads, epoch, batch, trainingDataPath, testingDataPath, model)
    

File: implementation/TrainNet.cpp
#include "Net.h"
#include "data_input.h"
#include <vector>
#include <cstdlib>


std::vector<double> forward(std::vector<double> val, Net neuralNet){
    std::vector<double> ret = neuralNet.ff(val, 0);
    ret = neuralNet.Sigmoid(ret);
    ret = neuralNet.ff(ret, 1);
    ret = neuralNet.Sigmoid(ret);
    ret = neuralNet.ff(ret, 2);
    // ret = neuralNet.Sigmoid(ret);
    // ret = neuralNet.ff(ret, 3);
    // ret = neuralNet.Sigmoid(ret);
    // ret = neuralNet.ff(ret, 4);
    ret = neuralNet.softMax(ret);
    return ret;
}

std::vector<std::vector<double>> oneHotEncode(std::vector<int> labels, int num_of_classes){
    std::vector<std::vector<double>> zeroed(labels.size(), std::vector<double>(num_of_classes, 0));
    for(int i = 0; i < labels.size(); i++){
        int label_val = labels.at(i);
        zeroed[i][label_val] = 1;
    }
    return zeroed;
}

int main(){
    std::cout << "start of Main\n";
    float learningRate = 0.02;
    int numThreads = 1;
    int epoch = 300;
    int batch = 8;

    std::string trainingDataPath = "mnist_train.csv";
    std::string testingDataPath = "winequality-red.csv";

    std::cout << "getting to this point.\n";

    data_input dataReader = data_input(trainingDataPath, ",");
    dataframe data = dataReader.getData(0);
    std::cout << "opened data\n";

    Net model = Net();
    model.addLayer(784, 128);
    model.addLayer(128, 64);
    model.addLayer(64, 10);

    model.setLearningRate(learningRate);

    // model.printDim(0);
    // model.printDim(1);
    // model.printDim(2);
    // model.printDim(3);
    // model.printDim(4);
    std::cout << "Set initial values\n";

    for(int i = 0; i < epoch; i++){
        std::cout << "epoch: " << i << "\n";

        dataframe shuffled = data.getSample(0.1);
        std::vector<int> truth = shuffled.getLabels();
        std::vector<std::vector<double>> train = shuffled.getData();
        std::vector<std::vector<double>> encoded_truth = oneHotEncode(truth, 10);

        int max_examples = encoded_truth.size();
        int index = 0;
        double avgLoss = 0;

        while(index < max_examples){
            model.zeroGrad();
            int counter = 0;

            while(counter < batch){
                if(index >= max_examples){
                    break;
                }
                std::cout << "in the while loop";

                std::vector<double> predictions = forward(train[index], model);
                double loss = model.crossEntropy(predictions, encoded_truth[index]);
                avgLoss += loss;
                model.backwardStep(predictions, encoded_truth[index]);
                index += 1;
                counter += 1;

                // model.printGrad();
            }

            model.updateWeights();
        }
        std::cout << "Avg Loss: " << avgLoss/max_examples << " for epoch: " << i << "\n";
    }

    return 0;
}
